### ISSUE TITLE
Non‑void returning assertion functions

### ISSUE THREAD

**Issue by ExE-Boss** [microsoft/TypeScript#40562]
## Search Terms

- assertion function non-void return type
- assertion function non-void return
- assertion function generic return
- assertion function custom return

## Suggestion

A way to type a function that is both an assertion function and returning a value that is not void.

## Use Cases

This is necessary to correctly type **Jest**’s `expect(…)` matchers, which have a generic return type of `R`.

## Examples

```ts
declare const expect: <T>(actual: T): JestMatchers<T>;

type JestMatchers<T> = JestMatchersShape<
	Matchers<void, T>,
	Matchers<
		Promise<void>,
		T extends PromiseLike<infer U>
			? U
			: Exclude<T, PromiseLike<any>>
	>
>;

type JestMatchersShape<TNonPromise extends {} = {}, TPromise extends {} = {}> = {
	resolves: AndNot<TPromise>;
	rejects: AndNot<TPromise>;
} & AndNot<TNonPromise>

type AndNot<T> = T & { not: T };

interface Matchers<R, T = {}> {
	toBe<E>(expected: E): R & (asserts T is E);
}

declare const foo: unknown;
expect(foo).toBe("foo");
foo; // $ExpectType "foo"
```

```ts
// Some typings for engine262's Completion Record handling:

export declare function ReturnIfAbrupt<T>(argument: T):
	(T extends NormalCompletion<infer V> ? V : Exclude<T, Completion>)
	& (asserts argument is T extends NormalCompletion<infer V> ? V : Exclude<T, Completion>);
export { ReturnIfAbrupt as Q };

/**
 * The type signature is the same, but `AssertNormalCompletion` causes an error to be thrown at runtime
 * if `argument` is an AbruptCompletion, whereas `ReturnIfAbrupt` gets replaced with code that causes
 * the caller to return the AbruptCompletion by engine262's build system:
 */
declare function AssertNormalCompletion<T>(argument: T):
	(T extends NormalCompletion<infer V> ? V : Exclude<T, Completion>)
	& (asserts argument is T extends NormalCompletion<infer V> ? V : Exclude<T, Completion>);
export { AssertNormalCompletion as X };
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by jcalz** [comment]
duplicate of or related to #34636

**Comment by ttencate** [comment]
Note that the duplicate #34636 (which got here first) has 39 :+1: at the time of writing, and this one has only 9 :+1:. If these are used for prioritization, perhaps this one should be closed and the other should be reopened?

**Comment by RebeccaStevens** [comment]
Here's a simpler example that doesn't require any 3rd party libraries:

```ts
type ValidRawData = { foo: string; };
declare function assertIsValidRawData(value: unknown): asserts value is ValidRawData;

// Ideal return type would be some like:
// ParsedData & asserts data is ValidRawData
function parseData(data: unknown) {
    assertIsValidRawData(
        data !== null &&
        typeof data === "object" &&
        Object.hasOwn(data, "foo") &&
        typeof data.foo === "string"
    );
    ...
}

const rawData: unknown = {};
const data = parseData(rawData);
const foo = rawData.foo; // <== Not currently valid. `parseData` called above should be able to assert that rawData is of type ValidRawData.
```

[Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAaghgGwJYBMBKcDuAROw5QC8UA3gFBSVQBmA9rQFxQDOwATkgHYDmA3GQF9+oSFAAKcNswgpc+IlABEYSdNl44i-mRQQAxgknRqAV057gSWpyhxm0tsACSzeMnRY5cABQA3RCYQTGYA1py0mJwAlEx2DsDMUP4IgVBIiW6oGDga2gD0eVBOuohQbBDAJmw2ItCYtCYIKFAARtDMtAC20MghQWQF4qoyXlAAZLb2EI6JKBppGYhZnhpkpuaW1lAqUhBe3nP4wZxhEdGkFFRx086uSx45+N6XVFSHBACEhMScjQjjYxerygtVo1Cg7yI3yUtBaACt9MBFACga8APLwxEAOgAFnY0ZEDhoADRKOi0RRRFHAqig8HvLHkqHERSsDg8RRAqL8V5A8qVapKHZqLxaQRkMh6aysMorI5QULhSIKEhCSXS4AQ+bEYV7DTeNhyuDciWDAAGuq8ZqgekQCBkthatF87RxDSarWgcBa9pBtEm8RBeM1hseBHSUDBIPA0EyDy8WPVnBlTOIoYT5PyBWzOdzefzBcLReLhdNhQAKjGAMp6DhgTUoWgQZicADkmrxLqgZrxzAJnGttWaIAqpI60dESE1nTgvztIETugMRig3AQsNK5EoXGA02ocD00AxCIsAGENWwTBZaGwLq9e-2ADzl0kAaSgEAAHrvOChEmI2FoSBHBAV8IBAAA+Z4aVhE9gCYF9USgPoQCYV8uSYWDEQWKBy3GKA0H0G8UEfV9SUVM4IJ5KABHFMgvzAG9NVVXggA).

**Comment by natew** [comment]
One footgun this avoids is it can avoid unused variables:

```tsx
const val = api.value
ensureExists(val)
// if you never use val for rest of file, but linter is happy
```

```tsx
const val = ensureExists(api.value)
// if you never use val for rest of file, linter will complain / editor will show unused
```

**Comment by natew** [comment]
The whole point of my post is that if I never use val, my longer/editor doesn’t tell me because it thinks it’s being used. This can happen when you used it, clean up some code below it that removes all the references to it, for example.  You miss seeing its not necessary anymore because it tricks the linter. Whereas the assign pattern if I remove all references below I’ll clearly see it’s dangling.

**Comment by irwincollin** [comment]
Would love for this feature to exist. Not only would it be useful for some mutating functions like `Object.assign` and the jest functions above, but I think it also enables some novel compile-time safe patterns for mutable structures that aren't possible to represent currently.

For example, a state machine that can **assert** that a transition is changing the underlying object to a new state, **and** return information about the transition itself.

Imagine this example:
```typescript
myMachine = createMachine();

myMachine.stopExport(); // error: stopExport does not exist on type MachineReady

myMachine.startExport() // returns returns a promise to await the export AND asserts that myMachine is in a different state
myMachine.startExport() // error: startExport does not exist on MachineStarted
```

**Comment by ITenthusiasm** [comment]
I agree that this would be a helpful feature. In addition to the `Object.assign()` scenario that @irwincollin mentioned above, there may be situations where a developer wants to mutate an array in place in order to avoid the performance overhead that comes from methods like `Array.map`. (`Array.map` would consume additional memory unnecessarily since it always creates a new array, and this can be problematic if the array is very large.)

The problem with returning something like `T & U` (like `Object.assign()`) is that such a return type won't communicate that the original variable which was passed to the function was mutated. And the problem with `asserts` is that it can't be used to return the mutated object for convenience.

A syntax like
```ts
function mapArrayInPlace<T, U>(array: T[], func: (items: T) => U): mutates array to U[];
```

could be helpful. I would hope that this would be fairly simple since TypeScript already has a way to `assert` and a way to return generic types. We'd just be adding 2 new keywords, and it would look familiar to what we have in other situations:

```ts
function predicate(arg: unknown): arg is number;
function asserter(arg: unknown): asserts arg is number;
function mutator(arg: unknown): mutates arg to number;
```

**Comment by ITenthusiasm** [comment]
This seems to be related to #17181, though it comes from the opposite angle. A library that rigorously identifies its mutating functions in the way described above could satisfy some of the requirements of that issue if I'm understanding correctly. (I only skimmed the issue.)

**Comment by TheUnlocked** [comment]
Another thing this could allow is an approximation of an ownership system. For example, say I have a `parse` function that I want to have parse an object based on a schema, including things like validation and filling in default values. For performance reasons, in some cases I might want the implementation to mutate the input to avoid needing to copy data when the parsed object is only slightly different from the input, but that would mean that the original value is corrupted. Allowing a return value with assertions would allow `parse` to _consume_ the input, like so:

```ts
interface Parser {
    parse<T>(obj: unknown, schema: Schema<T>): T & asserts obj is never;
}

const raw = getInput();
raw.x // ok
const parsed = parser.parse(raw, schema);
raw.x // error
```

**Comment by mkantor** [comment]
This is also related to #22865.

**Comment by artur-klesun-st** [comment]
Another extremely big use case is that this would finally open the door for mutable object argument type, a concept from Rust that makes typescript type system so much inferior...
```ts
function popKey<T extends object, TKey extends keyof T>(
  obj: T, key: TKey
): T[TKey], asserts obj is Omit<T, TKey> {
  const value = obj[key];
  delete obj[key];
  return value;
}

const obj = { a: 1, b: 2 };
const b = popKey(obj, 'b');
// should be a type error since we informed the compiler that `b` key was removed
obj.b.toFixed(2);
```
I guess you can kind of emulate it with an extra output mutable variable, but it would be awfully awkward.
